quint/src/quintAnalyzer.ts

```diff
@@ -58,11 +58,59 @@ export class QuintAnalyzer {
     this.modeChecker = new ModeChecker()
   }
 
+  setTable(lookupTable: LookupTable) {
+    this.typeInferrer.setTable(lookupTable)
+    this.effectInferrer.setTable(lookupTable)
+  }
+
+  setState(analysisOutput: AnalysisOutput) {
+    this.typeInferrer.types = analysisOutput.types
+    this.effectInferrer.effects = analysisOutput.effects
+    this.modeChecker.suggestions = analysisOutput.modes
+  }
+
+  // I'm actually not using this right now. I'm not sure we need it, as new
+  // expressions should get new ids.
+  deleteResult(id: bigint) {
+    this.typeInferrer.types.delete(id)
+    this.typeInferrer.errors.delete(id)
+    this.effectInferrer.effects.delete(id)
+    this.effectInferrer.errors.delete(id)
+    this.modeChecker.suggestions.delete(id)
+    this.modeChecker.errors.delete(id)
+  }
+
   analyze(module: QuintModule): void {
-    const [typeErrMap, types] = this.typeInferrer.inferTypes(module)
-    const [effectErrMap, effects] = this.effectInferrer.inferEffects(module)
+    const [typeErrMap, types] = this.typeInferrer.inferTypes(module.defs)
+    const [effectErrMap, effects] = this.effectInferrer.inferEffects(module.defs)
+    const updatesErrMap = this.multipleUpdatesChecker.checkEffects([...effects.values()])
+    const [modeErrMap, modes] = this.modeChecker.checkModes(module.defs, effects)
+
+    const errorTrees = [...typeErrMap, ...effectErrMap]
+
+    // TODO: Type and effec checking should return QuintErrors instead of error trees
+    this.errors.push(
+      ...errorTrees.map(([id, err]): [bigint, QuintError] => {
+        return [id, { code: 'QNT000', message: errorTreeToString(err), data: { trace: err } }]
+      })
+    )
+
+    this.errors.push(...modeErrMap.entries(), ...updatesErrMap.entries())
+
+    // We assume that ids are unique across modules, and map merging can be done
+    // without collision checks
+    this.output = {
+      types: new Map([...this.output.types, ...types]),
+      effects: new Map([...this.output.effects, ...effects]),
+      modes: new Map([...this.output.modes, ...modes]),
+    }
+  }
```

Comment:
These lines are all unchanged, it's just a duplication since we now have both `analyze` and `analyzeDef`